### PR TITLE
✨ add token driver for addon registration

### PR DESCRIPTION
**Pull request by elgnay**
## Related issue(s)

Fixes #

## Summary by CodeRabbit

* **New Features**
  * Token-based addon authentication path added alongside CSR, with runtime selection and automatic fallback.
  * New addon token configuration to control token expiration.

* **Improvements**
  * Registration flow enhanced to choose auth method per-addon and reduce unnecessary restarts via improved config hashing.
  * New CLI/Helm flags to expose addon auth type and token expiration.

* **Tests**
  * Expanded unit and integration tests covering token/CSR flows, rotation, transitions, and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: pkg/registration/register/interface.go [hunk 1]
```diff
@@ -81,9 +81,31 @@ type RegisterDriver interface {
...
+// AddonRegistrationOption provides complete configuration for addon registration,
+// including authentication method and access to driver options.
+// Registration type (KubeClient vs CustomSigner) is determined from SecretOption.Signer.
+type AddonRegistrationOption interface {
+	// GetAddonKubeClientRegistrationAuth returns the authentication method for addons with registration type KubeClient.
+	// Possible values are "csr" (default) and "token".
+	GetAddonKubeClientRegistrationAuth() string
+
+	// GetCSROption returns the CSR driver option
+	GetCSROption() interface{}
```
Comment (qiujian16) [inline pkg/registration/register/interface.go]: why this is not typed.

File: pkg/registration/register/interface.go [hunk 2]
```diff
@@ -81,9 +81,31 @@ type RegisterDriver interface {
...
+
+	// GetTokenOption returns the token driver option
+	GetTokenOption() interface{}
```
Comment (qiujian16) [inline pkg/registration/register/interface.go]: ditto

File: pkg/registration/register/aws_irsa/aws_irsa.go [hunk 3]
```diff
@@ -107,9 +123,78 @@ func (c *AWSIRSADriver) BuildClients(_ context.Context, secretOption register.Se
...
+
 	return clients, nil
 }
 
+func (c *AWSIRSADriver) Fork(addonName string, option register.AddonRegistrationOption, secretOption register.SecretOption) (register.RegisterDriver, error) {
+	// Determine registration type from signer name
+	isKubeClientType := secretOption.Signer == certificatesv1.KubeAPIServerClientSignerName
+
+	// Check if we should use token-based authentication for KubeClient registration type
+	if isKubeClientType && option.GetAddonKubeClientRegistrationAuth() == "token" {
```
Comment (qiujian16) [inline pkg/registration/register/aws_irsa/aws_irsa.go]: this seems like a common part for each driver, can we have a common function to start tokenDriver?

File: pkg/registration/register/factory/options.go [hunk 4]
```diff
@@ -45,6 +65,18 @@ func (s *Options) Validate() error {
 	}
 }
 
+func (s *Options) GetAddonKubeClientRegistrationAuth() string {
+	return s.AddonKubeClientRegistrationAuth
+}
+
+func (s *Options) GetCSROption() interface{} {
+	return s.CSROption
```
Comment (qiujian16) [inline pkg/registration/register/factory/options.go:73]: why this has to be interface?

File: pkg/registration/register/interface.go (hunk 1 above)
Comment (elgnay) [reply pkg/registration/register/interface.go]: To avoid Circular Dependency Occurs.

The `csr` and `token` packages need to import the parent register package to use:
  - register.SecretOption
  - register.RegisterDriver interface
  - register.AddonRegistrationOption interface
  - register.AddOnClients
  - Utility functions like register.KubeConfigFromSecretOption()

  If `register/interface.go` imports `register/csr` or `register/token`, Go will detect the circular dependency and fail to compile.

File: pkg/registration/register/factory/options.go (hunk 4 above)
Comment (elgnay) [reply pkg/registration/register/factory/options.go:73]: factory.Options.GetCSROption() cannot return a concrete type because it implements the register.AddonRegistrationOption interface which requires interface{} return type.

File: pkg/registration/register/interface.go (hunk 1 above)
Comment (qiujian16) [reply pkg/registration/register/interface.go]: If so why the interface needs to return csrOption and tokenOption for any driver. If it has to be returned on interface level, I think the returned type should also be defined in interface.

Comment (elgnay) [reply pkg/registration/register/interface.go]: > If so why the interface needs to return csrOption and tokenOption for any driver.
The cluster driver can fork add-on drivers at runtime, so the add-on registration needs to carry options for both driver types. Because csrOption and tokenOption are driver-specific, it’s debatable whether they should be part of the interface.

File: pkg/registration/spoke/addon/registration_controller.go [hunk 5]
```diff
@@ -127,14 +132,27 @@ func (c *addOnRegistrationController) syncAddOn(ctx context.Context, syncCtx fac
 		return nil
 	}
 
-	cachedConfigs := c.addOnRegistrationConfigs[addOnName]
-	configs, err := getRegistrationConfigs(addOn)
+	// Ensure driver field is set for kubeClient type registrations
+	updated, err := c.ensureDriver(ctx, addOn)
+	if err != nil {
```
Comment (qiujian16) [inline pkg/registration/spoke/addon/registration_controller.go]: if updated || err != nil {
  return err
}

**Comment by qiujian16** [comment]
looks good in general

File: pkg/registration/register/aws_irsa/aws_irsa.go (hunk 3 above)
Comment (elgnay) [reply pkg/registration/register/aws_irsa/aws_irsa.go]: A common function provided.

File: pkg/registration/spoke/addon/registration_controller.go (hunk 5 above)
Comment (elgnay) [reply pkg/registration/spoke/addon/registration_controller.go]: Updated as suggested.
